src/ai/defaultai.cc

```diff
@@ -3334,77 +3337,229 @@ void DefaultAI::check_flag_distances(const Time& gametime) {
 
 // Dealing with diplomacy actions
 void DefaultAI::diplomacy_actions(const Time& gametime) {
+	// Make sure we use reasonably up to date stats
+	update_player_stat(gametime);
```

Comment:
I've just seen a stupid move by an AI because of stale stats in a test. :)

Reply:
```
[01:54:19.707 game] DEBUG: AI Diplomacy: For player(1), the player(2) has the diploscore: 45
[01:54:19.707 game] DEBUG: AI Diplomacy: For player(1), the player(3) has the diploscore: 27
[01:54:19.707 game] DEBUG: AI Diplomacy: For player(1), the player(4) has the diploscore: 53
[01:54:21.077 game] DEBUG: AI Diplomacy: Player(2) plans to accept the invitation of player (1) with priority 17
[01:54:21.077 game] DEBUG: AI Diplomacy: Player(2) accepts the invitation of player (1) with diploscore 23.
[01:54:28.277 game] DEBUG: AI Diplomacy: For player(3), the player(1) has the diploscore: 6
[01:54:28.277 game] DEBUG: AI Diplomacy: For player(3), the player(2) has the diploscore: -34
[01:54:28.277 game] DEBUG: AI Diplomacy: For player(3), the player(4) has the diploscore: 23
[01:54:30.977 game] DEBUG: AI Diplomacy: Player(1) plans to leave team with priority 0
[01:54:30.977 game] DEBUG: AI Diplomacy: Player(1) leaves team (1) as last one.
```

If I'm not mistaken, that's less than the minimum allowed update interval of 15 seconds, so the above still does not make it impossible. Extreme bad luck for AI player 1, extreme good luck for me to catch problem. ;)

I made it to at least use consistent data in [63bdc6a](https://github.com/widelands/widelands/pull/5814/commits/63bdc6ab842d5efcdf6378c71e143ba007136476). I haven't checked thoroughly the rest of the function for similar possibilities though.

Reply:
I am not too happy with calling this outside its own scheduler. How do you define stale stats? it is updated every 60 seconds anyway, So the maximum stale time can be 60 seconds., while diplomacy is checked at an interval of 90 to 119 seconds. If you really feel we should update the diplomacy score just before making the decisions then we should move this part to the diplomacy action part. Furthermore we already tried to only compare data that is from the same point in time.  
I think it is basically impossible to prevent the AI from doing weird things from time to time and a silly diplomacy move is no problem from my side as long as it does not happen too often. Remember the whole thing might be affected by the mutations applied at game startup anyway.

Reply:
> I am not too happy with calling this outside its own scheduler.

I wasn't either, but I searched for it and it turned out that 2 functions in `defaultai_warfare.cc` do it too. (`check_enemy_sites()` and `check_militarysites()`)
git blame says it's been so since 2017.

> How do you define stale stats? it is updated every 60 seconds anyway, So the maximum stale time can be 60 seconds.

If you have a look at some logs, you can see that it's actually 17 to 19 seconds between stat updates of a player, probably because of the above. (it has a guard to prevent running it more often than once per 15 seconds)

So we have a more general problem here, that is out of scope for AI Diplo.

> Edit: what we could do is to add the call to diplomacy actions to the update stats task, and add some small time jitter there (i.e. update interval between 50 and 70 seconds). and by this delete the diplomacy task from the scheduler

I've thought about similar possibilities too. I'm not sure about threading and such, maybe we should just schedule it (at the end of every second run) and the other 2 from there, instead directly calling? But the others are currently scheduled more often (19 and 15 secs). Can we increase that?

Reply:
> > I am not too happy with calling this outside its own scheduler.
> 
> I wasn't either, but I searched for it and it turned out that 2 functions in `defaultai_warfare.cc` do it too. (`check_enemy_sites()` and `check_militarysites()`) git blame says it's been so since 2017.      
    
I will have a look, that seems to be a bloody workaround then. 
> 
> > How do you define stale stats? it is updated every 60 seconds anyway, So the maximum stale time can be 60 seconds.
> 
> If you have a look at some logs, you can see that it's actually 17 to 19 seconds between stat updates of a player, probably because of the above. (it has a guard to prevent running it more often than once per 15 seconds)
> 
> So we have a more general problem here, that is out of scope for AI Diplo.
> 
> > Edit: what we could do is to add the call to diplomacy actions to the update stats task, and add some small time jitter there (i.e. update interval between 50 and 70 seconds). and by this delete the diplomacy task from the scheduler
> 
> I've thought about similar possibilities too. I'm not sure about threading and such, maybe we should just schedule it (at the end of every second run) and the other 2 from there, instead directly calling? But the others are currently scheduled more often (19 and 15 secs). Can we increase that?   
   
currently AI isn't threaded at all. i.e. all AI calcualtions run in the same thread as the complete Economy and logic code. Only the display code is threaded so far AFAIK. Therefore AI Code directly affects Performance of our code.   I will have a look in the 2 issues. However if they only call the function update_player_stat at such an interval anyway we could simply lengthen the update state scheduler interval and call update_player_stat() and diplomacy_actions() really in one task,